* Developer Notes for further improvement
** Selecting the medoid instead of randomly selecting a MGCs from each family

To calculate a medoid, you need a distance matrix. Then, sum each row
to obtain the row totals, and select the lowest total (argmin). This will be
the MGC that has the lowest distance to all other MGcs.

This could be obtained by taking each MGC in the GCF and then taking
out the information again from the ‘fastani.results’ file to create
this distance matrix for each GCF.

Then, compute the column or row sum and apply
numpy.argmin(distMatrix.sum(axis=0))

This will be a job that will take ~4h of coding to add. So, I think
this could be something for Hannah to possibly take a look into.
** Phylogeny tree of predicted MGCs

I think for the publication we should show in here which kind of
clusters have been predicted in a phylogeny or something just to give
an idea of which clusters we are trying to predict that might be more
overrepresented in healthy samples since the CR collection was
obtained from healthy donors




** Focus on actual MGCs
Instead of focussing on putative MGCs, which can be generated by
gutSMASH. Try to focus mainly on actual gene clusters that are proven
to be known. So that we can hypothesize a bit more on the role of the
metabolite in health/disease? 

This can be achieved by running the final version of gutSMASH on the
reference genome set, to obtain the most accurate predctions. Then,
select for MGCs that are non-putative and filter them out. 
